snowflake_blueprints/upload_file.py

```diff
@@ -65,23 +59,209 @@ def get_args():
+    return snowflake_results
+
+
+def upload_data_with_insert(source_full_path,
+                            table_name,
+                            insert_method,
+                            db_connection):
+    """
+    Upload the data using pandas.to_sql which creates multiple INSERT statements.
+    """
+    print('Attempting upload with insert method')
+    chunksize = 10000
+    for index, chunk in enumerate(
+            pd.read_csv(source_full_path, chunksize=chunksize)):
+
+        if insert_method == 'replace' and index > 0:
+            # First chunk replaces the table, the following chunks
+            # append to the end.
+            insert_method = 'append'
+
+        chunk.columns = map(lambda x: str(x).upper(), chunk.columns)
+        print(f'Uploading rows {(chunksize*index)+1}-{chunksize * (index+1)}')
+        chunk.to_sql(
+            table_name,
+            con=db_connection,
+            index=False,
+            if_exists=insert_method,
+            method='multi',  # Not using pd_writer method due to inability to create tables and consistency with old behavior
+            chunksize=10000)
+
+
+def upload_data(
+        source_full_path,
+        table_name,
+        insert_method,
+        db_connection):
+    """
+    Upload the data to Snowflake. Tries the PUT method first, then relies on INSERT as a backup.
+    """
+    # Try to use put method initially.
+    try:
+        snowflake_results = upload_data_with_put(source_full_path,
+                                                 table_name,
+                                                 insert_method,
+                                                 db_connection)
+        # Needed to prevent other try from running if this is successful.
+        return snowflake_results
+    except ProgrammingError as pg_e:
+        if 'This session does not have a current schema.' in str(pg_e):
+            print(f'The schema provided either does not exist or your user does not have access to it. If no schema was provided, no default schema exists.')
+            print(pg_e)
+            sys.exit(errors.EXIT_CODE_INVALID_SCHEMA)
+        if 'This session does not have a current database.' in str(pg_e):
+            print(
+                f'The database provided either does not exist or your user does not have access to it. If no database was provided, this user does not have a default warehouse.')
+            print(pg_e)
+            sys.exit(errors.EXIT_CODE_INVALID_DATABASE)
+        print(pg_e)
+    except BaseException as e:
+        if 'No such file or directory:' in str(e):
+            print(
+                f'The combination of folder name/file name that you provided ({source_full_path}) could not be found. Please check for typos and try again.')
+            print(e)
+            sys.exit(errors.EXIT_CODE_FILE_NOT_FOUND)
+        print('Put method failed.')
+        print(e)
+        pass
```

Comment:
Print the exception if you didn't find what you were looking for from the string doesn't make sense. Wouldn't you want to still print the exception but not continue processing?

Reply:
@ericelsken No. I want it to try uploading with the PUT logic. If this fails for an unknown reason, it gets logged and trys to use the INSERT method. If it fails for a known reason (File does not exist, invalid schema, invalid DB, etc.) then it should exit and fail. There's no sense in trying the INSERT method if you know an input is invalid.

This was the only way I could figure out how to effectively have a "backup" upload method. Let me know if you have a better suggestion.